contracts/src/finance/vesting_wallet.rs

```diff
@@ -0,0 +1,621 @@
+            // and the function would have returned earlier.
+            scaled_allocation / self.duration()
+        }
+    }
+}
+
+#[cfg(all(test, feature = "std"))]
+mod tests {
+    use alloy_primitives::{address, uint, Address, U256, U64};
+    use stylus_sdk::block;
+
+    use super::{IVestingWallet, VestingWallet};
+
+    const TOKEN: Address = address!("A11CEacF9aa32246d767FCCD72e02d6bCbcC375d");
+    const DURATION: u64 = 4 * 365 * 86400; // 4 years
+
+    fn start() -> U64 {
+        U64::from(block::timestamp() + 3600) // 1 hour
+    }
+
+    #[motsu::test]
+    fn reads_start(contract: VestingWallet) {
+        let start = start();
+        contract._start.set(start);
+        assert_eq!(U256::from(start), contract.start());
+    }
+
+    #[motsu::test]
+    fn reads_duration(contract: VestingWallet) {
+        contract._duration.set(U64::from(DURATION));
+        assert_eq!(U256::from(DURATION), contract.duration());
+    }
+
+    #[motsu::test]
+    fn reads_end(contract: VestingWallet) {
+        let start = start();
+        let duration = U64::from(DURATION);
+        contract._start.set(start);
+        contract._duration.set(duration);
+        assert_eq!(U256::from(start + duration), contract.end());
+    }
+
+    #[motsu::test]
+    fn reads_max_end(contract: VestingWallet) {
+        let start = U64::MAX;
+        let duration = U64::MAX;
+        contract._start.set(start);
+        contract._duration.set(duration);
+        assert_eq!(U256::from(start) + U256::from(duration), contract.end());
+    }
+
+    #[motsu::test]
+    fn reads_released_eth(contract: VestingWallet) {
+        let one = uint!(1_U256);
+        contract._released.set(one);
+        assert_eq!(one, contract.released_eth());
+    }
+
+    #[motsu::test]
+    fn reads_released_erc20(contract: VestingWallet) {
+        let one = uint!(1_U256);
+        contract._erc20_released.setter(TOKEN).set(one);
+        assert_eq!(one, contract.released_erc20(TOKEN));
+    }
+
+    #[motsu::test]
+    fn gets_vesting_schedule(contract: VestingWallet) {
+        let start = start();
+        let duration = U64::from(DURATION);
+        contract._start.set(start);
+        contract._duration.set(duration);
+
+        let one = uint!(1_U256);
+        let two = uint!(2_U256);
+
+        assert_eq!(
+            U256::ZERO,
+            contract.vesting_schedule(two, start - U64::from(1))
+        );
+        assert_eq!(
+            one,
+            contract.vesting_schedule(two, start + duration / U64::from(2))
+        );
+        assert_eq!(two, contract.vesting_schedule(two, start + duration));
+        assert_eq!(
+            two,
+            contract.vesting_schedule(two, start + duration + U64::from(1))
+        );
+    }
+
+    #[motsu::test]
+    fn gets_vesting_schedule_zero_duration(contract: VestingWallet) {
+        let start = start();
+        contract._start.set(start);
+        contract._duration.set(U64::ZERO);
```

Comment:
Nit: If you're using this initialization frequently, make sense to introduce some kind of `fn init(start: U64, duration: u64)` function just for unit tests